src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/Tensor.cs

```diff
@@ -454,26 +459,44 @@ public Tensor<T> this[Tensor<bool> filter]
         /// Forms a slice out of the given tensor
         /// </summary>
         /// <param name="start">The ranges for the slice</param>
-        /// <returns><see cref="Tensor{T}"/> as a copy of the provided ranges.</returns>
-        // REVIEW: CURRENTLY DOES A COPY.
+        /// <returns><see cref="Tensor{T}"/> without copying the provided ranges.</returns>
         public Tensor<T> Slice(params ReadOnlySpan<NRange> start)
         {
             if (start.Length != Lengths.Length)
                 throw new ArgumentOutOfRangeException(nameof(start), "Number of dimensions to slice does not equal the number of dimensions in the span");
 
-            TensorSpan<T> s = AsTensorSpan(start);
-            T[] values = _isPinned ? GC.AllocateArray<T>(checked((int)s.FlattenedLength), _isPinned) : (new T[s.FlattenedLength]);
-            var outTensor = new Tensor<T>(values, s.Lengths.ToArray(), _isPinned);
-            s.CopyTo(outTensor);
-            return outTensor;
+            scoped Span<nint> lengths = new nint[Rank];
+            scoped Span<nint> offsets = new nint[Rank];
+
+            for (int i = 0; i < start.Length; i++)
+            {
+                (offsets[i], lengths[i]) = start[i].GetOffsetAndLength(Lengths[i]);
+            }
+
+            // When we have an empty Tensor and someone wants to slice all of it, we should return an empty Tensor.
+            // FlattenedLength is computed everytime so using a local to cache the value.
+            nint flattenedLength = FlattenedLength;
+            int memoryOffset = 0;
+
+            if (flattenedLength != 0)
+            {
+                for (int i = 0; i < offsets.Length; i++)
+                {
+                    memoryOffset += (int)(Strides[i] * offsets[i]);
+                }
+            }
+
+            if ((memoryOffset >= _values.Length || memoryOffset < 0) && flattenedLength != 0)
```

Comment:
>(memoryOffset >= _values.Length || memoryOffset < 0 [](http://example.com/codeflow?start=16&length=51)

Is this condition even possible? considering we have correct stride and offsets values.

Reply:
I'm actually not sure. I thought it was, which is why I put it in, but I'm wondering if the check in the `GetOffsetAndLength` covers it? I'll leave it for now but I'll experiment a bit more on it.